server.go

```diff
@@ -715,10 +716,14 @@ func (s *Server) processPublish(cl *Client, pk packets.Packet) error {
 		pk.FixedHeader.Qos = s.Options.Capabilities.MaximumQos // [MQTT-3.2.2-9] Reduce Qos based on server max qos capability
 	}
 
-	if pkx, err := s.hooks.OnPublish(cl, pk); err == nil {
+	pkx, err := s.hooks.OnPublish(cl, pk)
+	if err == nil {
 		pk = pkx
 	} else if errors.Is(err, packets.ErrRejectPacket) {
 		return nil
+	} else if err != nil && errors.As(err, new(packets.Code)) && cl.Properties.ProtocolVersion == 5 && pk.FixedHeader.Qos > 0 {
+		_ = cl.WritePacket(s.buildAck(pk.PacketID, packets.Puback, 0, pk.Properties, err.(packets.Code))) // omit connection error in favour of existing packet err
```

Comment:
When return err on the next line, it will result in disconnect with the reason as well.

Reply:
This was intentional as we already disconnect the client on any invalid packet due to protocol violation - but now I'm wondering if it's the correct behaviour for acks. What do you think?

Reply:
Looking at the spec, only the following reason code can be return on PUBACK:
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901124

I would think for these reason codes, it should not result in a disconnect for V5.